stdlib/Distributed/src/remotecall.jl

```diff
@@ -69,10 +70,15 @@ function test_existing_ref(r::AbstractRemoteRef)
     found = getkey(client_refs, r, nothing)
     if found !== nothing
         @assert r.where > 0
-        if isa(r, Future) && found.v === nothing && r.v !== nothing
-            # we have recd the value from another source, probably a deserialized ref, send a del_client message
-            send_del_client(r)
-            found.v = r.v
+        if isa(r, Future)
+            # this is only for copying the reference from Future to RemoteRef (just created)
+            fv_cache = @atomic :acquire found.v
+            rv_cache = @atomic :monotonic r.v
+            if fv_cache === nothing && rv_cache !== nothing
+                # we have recd the value from another source, probably a deserialized ref, send a del_client message
+                send_del_client(r)
+                @atomic :release found.v = rv_cache
```

Comment:
If multiple tasks are deserializing the same `ref`, only one of them should "win" here and update `found`. We probably need to do this with `:sequentially_consistent` order also, since that seems likely to be our intended memory model for `put!` / `take!` on AbstractRemoteRefs

Reply:
addressed it with an `@atomicreplace` - seems to be more fitting as its sequentially consistent and also keeps our rule of always setting the cache once